### PR TITLE
click on the map to select all points upstream

### LINKED ISSUE #2: Allow users to select a watershed using the map
Travis suggests users might like to select a watershed using the map, instead of a dropdown menu where they have to know the name of their watershed.

I would imagine the Leaflet map toolkit has a way to do that built in, it seems like something a lot of people would want. We'd have a button on the map to trigger watershed selection mode, maybe, then display all the watersheds. After a user picked one, it would go back to displaying only a single watershed outline to help the user understand what data they were viewing now (plus that watershed'd upstream and downstream networks!)

### LINKED ISSUE #43: Front end sometimes passes WKT and sometimes geoJSON
According to specifications, the front end should pass the backend WKT for `overlap` parameters. In practice, the front end sometimes passes WKT and sometimes passes geoJSON. The backend doesn't currently check the geometry data it receives, just passes it along to the database, which can interpret either format, so everything is working, but we'd definitely like to follow the spec here and also verify inputs in the backend.

Fix the front end to always pass WKT, then add geometry verification code to the backend.

### LINKED ISSUE #49: Remove stream network visualization on map
Visualization is "busy" and doesn't directly correspond to any data we show.

### PR DESCRIPTION
**Pull request by corviday**
Users may select an area of interest using the circlemarker tool on the map and then clicking on the map; this will select the region upstream of the selected point.

This PR also removes drawing the upstream stream network on the map. We aren't get doing meaningful visualization with it yet, and people found it distractingly busy, unfortunately.

Demo [here](https://services.pacificclimate.org/dev/scip/app/). 

I will lint changed files after review.

resolve #49 
resolve #2
resolve #43

### CODE REVIEW

**Comment by corviday** [comment]
one issue I've run into testing just now is that it is possible to select an area with so many vertices that sending it to the PCEX backend results in a request too long to be legal.

For example, if you click on -122.46875+52.53125, the resulting [requests](https://services.pacificclimate.org/dev/pcex/api/data?ensemble_name=scip_files&model=CanESM2&variable=apf_flow&emission=historical,+rcp85&timescale=yearly&time=0&area=MULTIPOLYGON+(((-121.4375+52.125,-121.4375+52.0625,-121.5+52.0625,-121.5625+52.0625,-121.5625+52.125,-121.5625+52.1875,-121.625+52.1875,-121.625+52.25,-121.6875+52.25,-121.6875+52.3125,-121.75+52.3125,-121.8125+52.3125,-121.875+52.3125,-121.9375+52.3125,-122+52.3125,-122.0625+52.3125,-122.125+52.3125,-122.125+52.375,-122.125+52.4375,-122.125+52.5,-122.1875+52.5,-122.25+52.5,-122.3125+52.5,-122.3125+52.4375,-122.375+52.4375,-122.375+52.5,-122.375+52.5625,-122.4375+52.5625,-122.4375+52.5,-122.5+52.5,-122.5625+52.5,-122.5625+52.4375,-122.5625+52.375,-122.625+52.375,-122.625+52.3125,-122.6875+52.3125,-122.75+52.3125,-122.8125+52.3125,-122.8125+52.25,-122.875+52.25,-122.9375+52.25,-123+52.25,-123+52.3125,-123+52.375,-123.0625+52.375,-123.125+52.375,-123.125+52.3125,-123.1875+52.3125,-123.25+52.3125,-123.3125+52.3125,-123.3125+52.375,-123.375+52.375,-123.375+52.3125,-123.4375+52.3125,-123.5+52.3125,-123.5625+52.3125,-123.5625+52.25,-123.625+52.25,-123.6875+52.25,-123.6875+52.3125,-123.75+52.3125,-123.75+52.375,-123.6875+52.375,-123.6875+52.4375,-123.75+52.4375,-123.75+52.5,-123.8125+52.5,-123.875+52.5,-123.875+52.4375,-123.9375+52.4375,-123.9375+52.5,-123.9375+52.5625,-124+52.5625,-124+52.625,-124+52.6875,-124.0625+52.6875,-124.125+52.6875,-124.125+52.75,-124.1875+52.75,-124.25+52.75,-124.3125+52.75,-124.375+52.75,-124.4375+52.75,-124.5+52.75,-124.5625+52.75,-124.5625+52.8125,-124.625+52.8125,-124.6875+52.8125,-124.75+52.8125,-124.8125+52.8125,-124.8125+52.75,-124.8125+52.6875,-124.875+52.6875,-124.9375+52.6875,-125+52.6875,-125.0625+52.6875,-125.125+52.6875,-125.1875+52.6875,-125.25+52.6875,-125.3125+52.6875,-125.375+52.6875,-125.375+52.75,-125.375+52.8125,-125.4375+52.8125,-125.5+52.8125,-125.5+52.875,-125.5625+52.875,-125.625+52.875,-125.6875+52.875,-125.6875+52.9375,-125.75+52.9375,-125.8125+52.9375,-125.8125+53,-125.875+53,-125.875+53.0625,-125.875+53.125,-125.9375+53.125,-125.9375+53.1875,-126+53.1875,-126.0625+53.1875,-126.125+53.1875,-126.125+53.125,-126.1875+53.125,-126.25+53.125,-126.25+53.0625,-126.3125+53.0625,-126.375+53.0625,-126.4375+53.0625,-126.4375+53,-126.5+53,-126.5+52.9375,-126.5625+52.9375,-126.625+52.9375,-126.625+53,-126.6875+53,-126.6875+52.9375,-126.75+52.9375,-126.8125+52.9375,-126.8125+53,-126.8125+53.0625,-126.875+53.0625,-126.9375+53.0625,-127+53.0625,-127.0625+53.0625,-127.0625+53.125,-127.0625+53.1875,-127.125+53.1875,-127.1875+53.1875,-127.25+53.1875,-127.25+53.25,-127.3125+53.25,-127.3125+53.3125,-127.375+53.3125,-127.375+53.375,-127.4375+53.375,-127.4375+53.4375,-127.5+53.4375,-127.5625+53.4375,-127.625+53.4375,-127.625+53.5,-127.6875+53.5,-127.75+53.5,-127.8125+53.5,-127.8125+53.5625,-127.8125+53.625,-127.8125+53.6875,-127.75+53.6875,-127.6875+53.6875,-127.6875+53.75,-127.625+53.75,-127.5625+53.75,-127.5+53.75,-127.5+53.8125,-127.4375+53.8125,-127.375+53.8125,-127.3125+53.8125,-127.3125+53.875,-127.3125+53.9375,-127.3125+54,-127.25+54,-127.25+54.0625,-127.1875+54.0625,-127.125+54.0625,-127.125+54.125,-127.0625+54.125,-127+54.125,-126.9375+54.125,-126.875+54.125,-126.8125+54.125,-126.8125+54.0625,-126.75+54.0625,-126.6875+54.0625,-126.6875+54.125,-126.75+54.125,-126.75+54.1875,-126.8125+54.1875,-126.8125+54.25,-126.8125+54.3125,-126.75+54.3125,-126.6875+54.3125,-126.6875+54.25,-126.625+54.25,-126.5625+54.25,-126.5625+54.1875,-126.5+54.1875,-126.4375+54.1875,-126.375+54.1875,-126.3125+54.1875,-126.3125+54.125,-126.25+54.125,-126.25+54.1875,-126.25+54.25,-126.1875+54.25,-126.125+54.25,-126.125+54.1875,-126.0625+54.1875,-126.0625+54.25,-126.0625+54.3125,-126.125+54.3125,-126.125+54.375,-126.0625+54.375,-126.0625+54.4375,-126.0625+54.5,-126+54.5,-125.9375+54.5,-125.875+54.5,-125.8125+54.5,-125.8125+54.4375,-125.75+54.4375,-125.6875+54.4375,-125.6875+54.375,-125.625+54.375,-125.625+54.3125,-125.5625+54.3125,-125.5+54.3125,-125.4375+54.3125,-125.375+54.3125,-125.3125+54.3125,-125.25+54.3125,-125.1875+54.3125,-125.125+54.3125,-125.0625+54.3125,-125.0625+54.375,-125+54.375,-124.9375+54.375,-124.875+54.375,-124.875+54.4375,-124.8125+54.4375,-124.8125+54.5,-124.875+54.5,-124.9375+54.5,-125+54.5,-125+54.4375,-125.0625+54.4375,-125.125+54.4375,-125.125+54.5,-125.1875+54.5,-125.25+54.5,-125.3125+54.5,-125.375+54.5,-125.4375+54.5,-125.5+54.5,-125.5625+54.5,-125.5625+54.5625,-125.625+54.5625,-125.6875+54.5625,-125.75+54.5625,-125.8125+54.5625,-125.8125+54.625,-125.875+54.625,-125.875+54.6875,-125.875+54.75,-125.875+54.8125,-125.9375+54.8125,-125.9375+54.875,-126+54.875,-126+54.9375,-126.0625+54.9375,-126.125+54.9375,-126.125+55,-126.1875+55,-126.1875+55.0625,-126.25+55.0625,-126.25+55.125,-126.25+55.1875,-126.3125+55.1875,-126.3125+55.25,-126.375+55.25,-126.375+55.3125,-126.4375+55.3125,-126.4375+55.375,-126.4375+55.4375,-126.4375+55.5,-126.4375+55.5625,-126.5+55.5625,-126.5+55.625,-126.5625+55.625,-126.625+55.625,-126.6875+55.625,-126.6875+55.6875,-126.75+55.6875,-126.8125+55.6875,-126.8125+55.75,-126.8125+55.8125,-126.875+55.8125,-126.9375+55.8125,-126.9375+55.875,-126.9375+55.9375,-127+55.9375,-127+56,-127+56.0625,-127.0625+56.0625,-127.0625+56.125,-127.0625+56.1875,-127+56.1875,-126.9375+56.1875,-126.875+56.1875,-126.875+56.125,-126.8125+56.125,-126.8125+56.0625,-126.75+56.0625,-126.6875+56.0625,-126.6875+56.125,-126.625+56.125,-126.5625+56.125,-126.5+56.125,-126.4375+56.125,-126.375+56.125,-126.375+56.0625,-126.3125+56.0625,-126.25+56.0625,-126.25+56,-126.25+55.9375,-126.1875+55.9375,-126.1875+55.875,-126.125+55.875,-126.125+55.8125,-126.0625+55.8125,-126.0625+55.75,-126+55.75,-125.9375+55.75,-125.9375+55.6875,-125.9375+55.625,-125.875+55.625,-125.875+55.5625,-125.8125+55.5625,-125.8125+55.5,-125.8125+55.4375,-125.75+55.4375,-125.6875+55.4375,-125.625+55.4375,-125.625+55.375,-125.5625+55.375,-125.5625+55.3125,-125.5+55.3125,-125.5+55.25,-125.5+55.1875,-125.4375+55.1875,-125.4375+55.125,-125.375+55.125,-125.3125+55.125,-125.25+55.125,-125.1875+55.125,-125.125+55.125,-125.0625+55.125,-125.0625+55.0625,-125+55.0625,-124.9375+55.0625,-124.875+55.0625,-124.8125+55.0625,-124.75+55.0625,-124.75+55,-124.6875+55,-124.6875+55.0625,-124.625+55.0625,-124.5625+55.0625,-124.5+55.0625,-124.5+55,-124.4375+55,-124.375+55,-124.3125+55,-124.25+55,-124.1875+55,-124.125+55,-124.0625+55,-124.0625+55.0625,-124+55.0625,-124+55,-123.9375+55,-123.875+55,-123.8125+55,-123.75+55,-123.75+54.9375,-123.6875+54.9375,-123.6875+54.875,-123.6875+54.8125,-123.6875+54.75,-123.6875+54.6875,-123.625+54.6875,-123.5625+54.6875,-123.5625+54.625,-123.5+54.625,-123.5+54.5625,-123.4375+54.5625,-123.375+54.5625,-123.3125+54.5625,-123.25+54.5625,-123.1875+54.5625,-123.125+54.5625,-123.0625+54.5625,-123.0625+54.625,-123+54.625,-122.9375+54.625,-122.9375+54.5625,-122.875+54.5625,-122.875+54.5,-122.875+54.4375,-122.9375+54.4375,-122.9375+54.375,-122.875+54.375,-122.875+54.3125,-122.8125+54.3125,-122.8125+54.375,-122.75+54.375,-122.6875+54.375,-122.6875+54.3125,-122.625+54.3125,-122.5625+54.3125,-122.5625+54.375,-122.5+54.375,-122.4375+54.375,-122.375+54.375,-122.3125+54.375,-122.3125+54.4375,-122.25+54.4375,-122.25+54.5,-122.1875+54.5,-122.125+54.5,-122.0625+54.5,-122.0625+54.4375,-122+54.4375,-121.9375+54.4375,-121.875+54.4375,-121.8125+54.4375,-121.75+54.4375,-121.6875+54.4375,-121.625+54.4375,-121.5625+54.4375,-121.5+54.4375,-121.4375+54.4375,-121.4375+54.5,-121.4375+54.5625,-121.375+54.5625,-121.3125+54.5625,-121.25+54.5625,-121.1875+54.5625,-121.125+54.5625,-121.125+54.5,-121.125+54.4375,-121.0625+54.4375,-121+54.4375,-120.9375+54.4375,-120.875+54.4375,-120.875+54.5,-120.8125+54.5,-120.8125+54.4375,-120.75+54.4375,-120.75+54.375,-120.6875+54.375,-120.6875+54.3125,-120.625+54.3125,-120.625+54.25,-120.5625+54.25,-120.5625+54.1875,-120.5625+54.125,-120.5+54.125,-120.4375+54.125,-120.375+54.125,-120.3125+54.125,-120.25+54.125,-120.1875+54.125,-120.1875+54.0625,-120.25+54.0625,-120.25+54,-120.25+53.9375,-120.1875+53.9375,-120.125+53.9375,-120.125+54,-120.0625+54,-120.0625+53.9375,-120.0625+53.875,-120.0625+53.8125,-120+53.8125,-119.9375+53.8125,-119.9375+53.75,-119.875+53.75,-119.875+53.6875,-119.8125+53.6875,-119.75+53.6875,-119.75+53.625,-119.8125+53.625,-119.875+53.625,-119.875+53.5625,-119.875+53.5,-119.8125+53.5,-119.8125+53.4375,-119.75+53.4375,-119.75+53.375,-119.6875+53.375,-119.625+53.375,-119.5625+53.375,-119.5+53.375,-119.4375+53.375,-119.375+53.375,-119.375+53.3125,-119.3125+53.3125,-119.3125+53.25,-119.3125+53.1875,-119.25+53.1875,-119.1875+53.1875,-119.125+53.1875,-119.0625+53.1875,-119.0625+53.125,-119+53.125,-119+53.1875,-119+53.25,-118.9375+53.25,-118.875+53.25,-118.8125+53.25,-118.8125+53.1875,-118.75+53.1875,-118.75+53.125,-118.6875+53.125,-118.6875+53.0625,-118.625+53.0625,-118.625+53,-118.5625+53,-118.5625+52.9375,-118.5+52.9375,-118.4375+52.9375,-118.4375+52.875,-118.375+52.875,-118.375+52.8125,-118.375+52.75,-118.3125+52.75,-118.25+52.75,-118.25+52.6875,-118.25+52.625,-118.25+52.5625,-118.25+52.5,-118.3125+52.5,-118.375+52.5,-118.4375+52.5,-118.5+52.5,-118.5625+52.5,-118.5625+52.5625,-118.5625+52.625,-118.625+52.625,-118.6875+52.625,-118.75+52.625,-118.75+52.6875,-118.8125+52.6875,-118.8125+52.75,-118.875+52.75,-118.9375+52.75,-118.9375+52.8125,-119+52.8125,-119.0625+52.8125,-119.125+52.8125,-119.1875+52.8125,-119.1875+52.75,-119.25+52.75,-119.3125+52.75,-119.375+52.75,-119.4375+52.75,-119.5+52.75,-119.5625+52.75,-119.625+52.75,-119.6875+52.75,-119.6875+52.6875,-119.75+52.6875,-119.75+52.625,-119.8125+52.625,-119.8125+52.6875,-119.875+52.6875,-119.875+52.75,-119.9375+52.75,-120+52.75,-120+52.8125,-120.0625+52.8125,-120.0625+52.875,-120.125+52.875,-120.1875+52.875,-120.25+52.875,-120.3125+52.875,-120.3125+52.8125,-120.3125+52.75,-120.3125+52.6875,-120.25+52.6875,-120.25+52.625,-120.3125+52.625,-120.3125+52.5625,-120.375+52.5625,-120.375+52.5,-120.4375+52.5,-120.5+52.5,-120.5+52.4375,-120.5+52.375,-120.5625+52.375,-120.5625+52.3125,-120.5+52.3125,-120.5+52.25,-120.4375+52.25,-120.375+52.25,-120.375+52.1875,-120.375+52.125,-120.4375+52.125,-120.4375+52.1875,-120.5+52.1875,-120.5625+52.1875,-120.625+52.1875,-120.625+52.125,-120.625+52.0625,-120.6875+52.0625,-120.6875+52.125,-120.75+52.125,-120.8125+52.125,-120.875+52.125,-120.9375+52.125,-121+52.125,-121.0625+52.125,-121.125+52.125,-121.1875+52.125,-121.25+52.125,-121.3125+52.125,-121.375+52.125,-121.4375+52.125)))) to the backend return the error message 

```
Bad Request
Request Line is too large (8192 > 4094) 
```

Possible fixes:

1. change gunicorn's request line size - the best option if it works, but the backend may not be able to handle a shape this complex either
2. just tell the user to pick another region, that one is too complex
3. shorten the string in the front end some easy way, perhaps by rounding all the decimals
4. shorten the string in the front end some better way, by using some kind of simplification algorithm

**Comment by corviday** [comment]
The PCEX backend has been modified to accept extra-long polygons as request paylods. Just need to update this front end.

And maybe the SCIP backend as well?

**Comment by corviday** [comment]
Searching for a request-making library that will let me send a `GET` request with a body.

- axios doesn't raise an error but doesn't send the body
- fetch raises an error if you try to send a `GET` with a body

**Comment by corviday** [comment]
No library seems to be able to make a `GET` request with a body; possibly browsers do not allow it. Accordingly, have shifted tactics to accept` POST` requests instead.

**Comment by corviday** [comment]
[Demo](https://services.pacificclimate.org/dev/scip/app/)

**Comment by rod-glover** [comment]
Before I review further, a question: Your PR and the in-app instructions say "place a marker and select everything _upstream_ of it", but it looks to me like the map displays the downstream path. Examples:

Screenshot from 2023-11-15 14-47-38
Screenshot from 2023-11-15 14-49-11
Screenshot from 2023-11-15 14-49-42

Can you clarify?

**Comment by corviday** [comment]
The area of interest, whether selected from a dropdown or by clicking on the map, is outlined in white. It's pretty small on your selection because most points on the map have only a couple grid squares upstream, if there's even anything upstream of them.

The blue outline shows the downstream path to the sea. Salmon scientists have indicated this is a useful thing to show on the map, though as of yet we are not offering any indicator data long this route. Hope to someday.

**Comment by rod-glover** [comment]
Ah, I see. Thanks. I wonder if a different colour would lower this confusion. Also, perhaps instead, in the instructions, add a note that the map shows (a) the downstream path in blue, and (b) the upstream polygon in white|whatever.

**Comment by corviday** [comment]
I am wondering if it might help to make the circle marker where you click white (to match the boundary of the watershed you've chosen) instead of the default blue (which matches the downstream line). That way there would be a visual connection between the clicked location and the watershed.

**Comment by rod-glover** [comment]
That would help, too. I do also recommend making the "instructions" below the map more explicit with something like what I suggested.

**Comment by QSparks** [comment]
The demo looks great Lee! I like that the marker size changes with the zoom level.

During testing, I discovered a bug while using the `Edit layers` option. This option allows a marker that has been placed on the map to be moved without the changes being reflected in the upstream selection.

An additional suggestion for clarity is that for the `Delete layers` option, it looks like both “save” and “clear all” have the same behavior. If saving isn’t important here the options could be reduced to either clear all or cancel.

**Comment by corviday** [comment]
UI Improvements: Removed the "edit layers" button. After research, I am not sure how to get rid of the "save" menu option, so I might kick that to a separate issue. Added text explaining that the selected area is outlined in white.

File: src/data-services/pcex-backend.js [hunk 1]
```diff
@@ -53,42 +53,72 @@ export function flattenMultimeta(response) {
...
+    if (area && wkt.length > MAXAREALENGTH) {
+        // send a POST reqiest to accomodate long area string
+        const formData = new FormData();
+        formData.append('id_', datafile);
+        formData.append('area', wkt);
+        formData.append('variable', variable);
+        return axios.post(
+            `${process.env.REACT_APP_PCEX_API_URL}/timeseries`,
+            formData
+        ).then(response => response.data);
```
Comment (rod-glover) [inline src/data-services/pcex-backend.js:70]: This is the magic!

File: src/data-services/scip-backend.js [hunk 2]
```diff
@@ -1,68 +1,89 @@
...
+            params.common_name = common_name;
+            if(subgroup) {
+                params.subgroup = subgroup;
+            }
+        }
+        return axios.get(
+            `${process.env.REACT_APP_SCIP_API_URL}/region`, { params: params}
+        ).then(response=> response.data);
     }
+}
```
Comment (rod-glover) [inline src/data-services/scip-backend.js:75]: The pattern of sending a request as either a GET or a POST is common to 4 or 5 data services. I wonder, but am not sure, if it makes sense to factor that out as a single utility function and just pass in the details to it.

**Review by rod-glover — APPROVED** [review]
Code looks good. (Maybe DRY up GET vs POST request sends?)

In the latest demo, the basins(?) don't show up, whereas before IIRC they were filled in purple. Was that intentional?

Just one comment to repeat yet again ... to my eye, the white colour for the upstream polygon boundary is hard to see, particularly without the colour fill on the basin polygons for contrast, and particularly when the upstream region is small. I think a black boundary would be easier to see.

**Comment by corviday** [comment]
> In the latest demo, the basins(?) don't show up, whereas before IIRC they were filled in purple. Was that intentional?

Yes. We were requested to [remove the upstream flow network visualization](https://github.com/pacificclimate/scip-frontend/issues/49), which was a collection of lines that, when zoomed out, made the whole basin look blue. #

**Pull request merged**